File: benchmarks/tsb/bench_interpolate_zero_nearest.ts
/**
 * Benchmark: interpolateSeries with zero and nearest methods.
 * Outputs JSON: {"function": "interpolate_zero_nearest", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, interpolateSeries } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

// ~15% null values with consecutive gaps
const data: (number | null)[] = Array.from({ length: SIZE }, (_, i) => {
  const mod = i % 7;
  if (mod === 0 || mod === 1 || mod === 2) return null;
  return Math.sin(i * 0.01) * 100;
});
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  interpolateSeries(s, { method: "zero" });
  interpolateSeries(s, { method: "nearest" });
  interpolateSeries(s, { method: "linear", limit: 2 });
  interpolateSeries(s, { method: "ffill", limit: 5 });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  interpolateSeries(s, { method: "zero" });
  interpolateSeries(s, { method: "nearest" });
  interpolateSeries(s, { method: "linear", limit: 2 });
  interpolateSeries(s, { method: "ffill", limit: 5 });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "interpolate_zero_nearest",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_interpolate_zero_nearest.py
"""Benchmark: Series.interpolate with zero and nearest methods."""
import json, time
import numpy as np
import pandas as pd

SIZE = 50_000
WARMUP = 5
ITERATIONS = 30

data = [None if i % 7 in (0, 1, 2) else np.sin(i * 0.01) * 100 for i in range(SIZE)]
s = pd.Series(data, dtype="float64")

for _ in range(WARMUP):
    s.interpolate(method="zero")
    s.interpolate(method="nearest")
    s.interpolate(method="linear", limit=2)
    s.interpolate(method="pad", limit=5)

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.interpolate(method="zero")
    s.interpolate(method="nearest")
    s.interpolate(method="linear", limit=2)
    s.interpolate(method="pad", limit=5)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "interpolate_zero_nearest",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
